src/py2spack/loading.py

```diff
@@ -0,0 +1,266 @@
+"""Utils for downloading packages from PyPI and opening the pyproject.tomls."""
+
+import io
+import tarfile
+from typing import Any, Dict, List, Optional
+
+import requests  # type: ignore
+import tomli
+from packaging import version as vn
+
+KNOWN_ARCHIVE_FORMATS = [
+    ".tar",
+    ".tar.gz",
+    ".tar.bz2",
+    ".gz",
+    ".xz",
+    ".bz2",
+]
+
+
+class APIError(Exception):
+    """Error during version lookup or querying of the PyPI JSON API."""
+
+    def __init__(
+        self,
+        msg: str,
+    ):
+        """Initialize error."""
+        super().__init__(msg)
+
+
+def _acceptable_version(version: str) -> Optional[vn.Version]:
+    """Check whether version string can be parsed and has correct format."""
+    try:
+        v = vn.parse(version)
+        # do not support post releases of prereleases etc.
+        if v.pre and (v.post or v.dev or v.local):
+            return None
+        return v
+    except vn.InvalidVersion:
+        return None
+
+
+class PyPILookup:
```

Comment:
@egparedes Regarding the `get_files` interface: 
The method currently only provides metadata of the distribution package like filename, download url, hash, etc.
But it would make sense to directly put methods for loading the pyproject.toml file contents etc. here too. And I agree, this could be done for a specific version of the package. 
I think a public method to get the `pyproject` should be enough though (no public method to get the actual distribution package), since we only need the distribution package to get the pyproject.toml data. So this can just be downloaded and extracted in a helper method.

So in summary, my idea would be to do a protocol with public methods `get_versions(name)` which returns a list of available versions (already filtered for only source distributions?) and `get_pyproject(name, version)` which returns a dict with the contents of the pyproject.toml file.

If later we also need to get other files from the source distribution except the `pyproject.toml`, I could also make a more generic method to extract a file from the source distribution?

Reply:
Yes, I agree. Let's keep it simple for now.